### PR TITLE
[TechDocs] Remove deprecated preparer support and document ref values

### LINKED ISSUE #4409: TechDocs: Deprecate legacy `git clone` and dir preparer in favor of URL Reader
# What?

When [TechDocs is set to generate docs](https://backstage.io/docs/features/techdocs/getting-started#should-techdocs-backend-generate-docs) on the Backstage server, it has to "prepare" the source files before it can generate the site. In the early days, TechDocs implemented a simple `git clone` way of downloading the repository. If your [`backstage.io/techdocs-ref`](https://backstage.io/docs/features/software-catalog/well-known-annotations#backstageiomanaged-by-origin-location) annotation is not prefixed with `url:`, but has `github:` or `gitlab:`, TechDocs will use the legacy `git clone` preparer. The `dir:` is a syntactic sugar for the git preparer.

Later on, URL Reader received the feature of `readTree` by which it can read files and trees by downloading an archive of the repository.

# Why deprecate?

1. URL Reader (downloading archive) is much faster than Common git preparer (a git clone). TechDocs preparer step only needs markdown files and not the git history of the repository, which could increase the size 10x for a repository as large as Backstage mono-repo (Prepare times reduces from 1 minute to 5 seconds)
2. URL Reader is used across Backstage and not just TechDocs. Thus, it is much actively maintained and receives frequent support. Common Git Preparers are maintained by TechDocs and only used in it.
3. URL Reader supports more source code host providers (GitHub/GHE/GitLab(and hosted)/BitBucket(and hosted)/Azure DevOps (and on-premises)) and their intricacies like different API versions than Common Git Preparers in TechDocs. Using URL Reader adds support for GHE, Bitbucket, etc. It's quite likely that URL Reader will continue to be stable and exhaustive with more providers.
4. `dir:` preparer is complicated to maintain alongside `url:` preparer (and is only used and maintained by TechDocs). I feel that the syntactic sugar is not worth the effort of maintaining it for just one plugin. `url:` is _standard_ across backstage.

# How to migrate?

The [software templates](https://github.com/backstage/backstage/blob/a034569199aa59830aad18e2b6ea91253d8dbc70/plugins/scaffolder-backend/sample-templates/docs-template/%7B%7Bcookiecutter.component_id%7D%7D/catalog-info.yaml#L8) in this mono-repo already uses the `url:` format in `backstage.io/techdocs-ref` annotation in the Entity's `catalog-info.yaml`.

If you have created your own Software Templates, which use `github:`, `gitlab:`, `dir:`, etc. for `backstage.io/techdocs-ref` annotation. - update them with `url:` format. You also need to update existing `catalog-info.yaml` that have been created and use TechDocs annotation. Here is a simple HOW TO guide to help you https://backstage.io/docs/features/techdocs/how-to-guides#how-to-use-url-reader-in-techdocs-prepare-step

Note that `dir:` was sometimes used for local development when working with TechDocs. Now, we can use https://github.com/backstage/techdocs-cli for preview when writing docs locally and to debug any build errors. However, an alternative will be to keep dir preparer but not maintain it for production use-case.

# Plan

Throw warning in backend logs when non-URL preparer is used, with the migration link. Release on Day X.

In the release after X + 30 days, drop support for git and dir preparers in TechDocs.

Open for comments.

### PR DESCRIPTION
**Pull request by iamEAP**
## Hey, I just made a Pull Request!

Follow-up to @dhenneke's awesome work in #6495, which:

- Documents the recommended `backstage.io/techdocs-ref` annotation values
- Updates all default values in templates to use `dir:.`
- Fully removes support for deprecated `github:` / `gitlab:` / etc. annotation values
- Fully removes all code that went into supporting the deprecated git/common preparer(s).

Closes #4409

#### :heavy_check_mark: Checklist

- [x] A changeset describing the change and affected packages. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#creating-changesets))
- [x] Added or updated documentation
- [ ] Tests for new functionality and regression tests for bug fixes
- [ ] Screenshots attached (for UI changes)
- [x] All your commits have a `Signed-off-by` line in the message. ([more info](https://github.com/backstage/backstage/blob/master/CONTRIBUTING.md#developer-certificate-of-origin))

### CODE REVIEW

**Review by camilaibs — APPROVED** [review]

File: packages/techdocs-common/src/stages/generate/helpers.test.ts [hunk 1]
```diff
@@ -147,7 +147,7 @@ describe('helpers', () => {
 
     it('should return correct target url for supported hosts', () => {
       const parsedLocationAnnotation1: ParsedLocationAnnotation = {
-        type: 'github',
+        type: 'github' as RemoteProtocol,
```
Comment (dhenneke) [inline packages/techdocs-common/src/stages/generate/helpers.test.ts:150]: `getRepoUrlFromLocationAnnotation` isn't very helpful in its current state as it only supports the deleted `github` and `gitlab` types. But it should be possible to use the `ScmIntegrations` to figure out which kind of integration it is instead. But that's probably a task for a dedicated PR.

**Review by dhenneke — APPROVED** [review]
Thanks for the cleanup and the docs update 👍

File: plugins/scaffolder-backend/sample-templates/create-react-app/skeleton/catalog-info.yaml [hunk 2]
```diff
@@ -5,7 +5,7 @@ metadata:
   description: {{cookiecutter.description | jsonify}}
   annotations:
     github.com/project-slug: {{cookiecutter.destination.owner + "/" + cookiecutter.destination.repo}}
-    backstage.io/techdocs-ref: url:{{"https://" + cookiecutter.destination.host + "/" + cookiecutter.destination.owner + "/" + cookiecutter.destination.repo}}
+    backstage.io/techdocs-ref: dir:.
```
Comment (OrkoHunter) [inline plugins/scaffolder-backend/sample-templates/create-react-app/skeleton/catalog-info.yaml:8]: Oh, look at this gem. 💎

Comment (OrkoHunter) [reply plugins/scaffolder-backend/sample-templates/create-react-app/skeleton/catalog-info.yaml:8]: We should update https://github.com/backstage/software-templates as well :)

**Review by OrkoHunter — APPROVED** [review]
🚀 🌔

File: packages/techdocs-common/src/stages/generate/helpers.test.ts (hunk 1 above)
Comment (iamEAP) [reply packages/techdocs-common/src/stages/generate/helpers.test.ts:150]: Yeah, will create a follow-up issue to track.

**Pull request merged**